traversal/procedure/GraphProcedure.java

```diff
@@ -176,10 +176,17 @@ private void assertWithinFilterBounds(Set<Identifier.Variable.Retrievable> filte
             LOG.trace(this.toString());
         }
         assertWithinFilterBounds(filter);
-        return async(startVertex().iterator(graphMgr, params).map(
-                // TODO we can reduce the size of the distinct() set if the traversal engine doesn't overgenerate as much
-                v -> new GraphIterator(graphMgr, v, this, params, filter).distinct()
-        ), parallelisation);
+        if (startVertex().id().isRetrievable() && filter.contains(startVertex().id().asVariable().asRetrievable())) {
+            return async(startVertex().iterator(graphMgr, params).map(
+                    // TODO we can reduce the size of the distinct() set if the traversal engine doesn't overgenerate as much
+                    v -> new GraphIterator(graphMgr, v, this, params, filter).distinct()
+            ), parallelisation);
+        } else {
```

Comment:
if the start vertex is NOT part of the filtered/returned set of variables, we need to put the `distinct()` set at the outer layer, or we can have answers from remainder of the variables

we do this for both the async producer and non-parallel iterator methods